### PR TITLE
[Unicode API] Add wide char for compiler model APIs

### PR DESCRIPTION
**Pull request by zhaixuejun1993**
### Details:
 - Add API: CompiledModel compile_model(const std::wstring& model_path, const AnyMap& properties = {});
 - Add API: util::EnableIfAllStringAny<CompiledModel, Properties...> compile_model(const std::wstring& model_path, Properties&&... properties);
 - Add API: CompiledModel compile_model(const std::wstring& model_path, const std::string& device_name, const AnyMap& properties = {});
- Add API: util::EnableIfAllStringAny<CompiledModel, Properties...> compile_model(const std::wstring& model_path, const std::string& device_name, Properties&&... properties)
### Tickets:
 - 25101

### CODE REVIEW

**Review by ilyachur — COMMENTED** [review]
Can we add new tests for this API?

**Comment by zhaixuejun1993** [comment]
> Can we add new tests for this API?

@ilyachur Yes, of cause, will create the unite test for those APIs. To keep the same as other UNICODE APIs, I need a reference, where can I find the test for this Unicode API:
#ifdef OPENVINO_ENABLE_UNICODE_PATH_SUPPORT
    /**
     * @brief Reads models from IR / ONNX / PDPD / TF / TFLite file formats.
     * @param model_path Path to a model.
     * @param bin_path Path to a data file.
     * For IR format (*.bin):
     *  * if `bin_path` is empty, will try to read a bin file with the same name as xml and
     *  * if the bin file with the same name is not found, will load IR without weights.
     * For the following file formats the `bin_path` parameter is not used:
     *  * ONNX format (*.onnx)
     *  * PDPD (*.pdmodel)
     *  * TF (*.pb)
     *  * TFLite (*.tflite)
     * @return A model.
     */
    std::shared_ptr<ov::Model> read_model(const std::wstring& model_path, const std::wstring& bin_path = {}) const;
#endif

File: src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp [hunk 1]
```diff
@@ -203,6 +204,60 @@ TEST(OVClassBasicTest, smoke_createMockEngineConfigThrows) {
...
+                       << ::ov::util::wstring_to_string(modelXmlPathW) << "'";
+            }
+
+            is_copy_successfully = CommonTestUtils::copyFile(model_bin_name, modelBinPathW);
+            if (!is_copy_successfully) {
+                FAIL() << "Unable to copy from '" << model_bin_name << "' to '"
+                       << ::ov::util::wstring_to_string(modelBinPathW) << "'";
+            }
+
+            GTEST_COUT << "Test " << testIndex << std::endl;
```
Comment (riverlijunjie) [inline src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp:238]: Do we need this log?

File: src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp [hunk 2]
```diff
@@ -203,6 +204,60 @@ TEST(OVClassBasicTest, smoke_createMockEngineConfigThrows) {
...
+
+            ov::Core core = createCoreWithTemplate();
+            GTEST_COUT << "Core created " << testIndex << std::endl;
+
+            OV_ASSERT_NO_THROW(core.compile_model(modelXmlPathW));
```
Comment (riverlijunjie) [inline src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp:243]: We also need add tests for other new unicode APIs, right?

Comment (zhaixuejun1993) [reply src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp:243]: > We also need add tests for other new unicode APIs, right?

Yes, the other will be update later. This is for quickly look to make sure add test here is right.

File: src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp (hunk 1 above)
Comment (zhaixuejun1993) [reply src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp]: removed

File: src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp [hunk 3]
```diff
@@ -203,6 +204,185 @@ TEST(OVClassBasicTest, smoke_createMockEngineConfigThrows) {
...
+            }
+
+            ov::Core core = createCoreWithTemplate();
+
+            OV_ASSERT_NO_THROW(core.compile_model(modelXmlPathW, CommonTestUtils::DEVICE_CPU,
+                               ov::hint::performance_mode(ov::hint::PerformanceMode::LATENCY)));
+            CommonTestUtils::removeFile(modelXmlPathW);
+            CommonTestUtils::removeFile(modelBinPathW);
+            GTEST_COUT << "OK" << std::endl;
+        } catch (const ov::Exception& e_next) {
```
Comment (ilyachur) [inline src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp:372]: Why do we need this `try {} catch {}`? Looks like  we use `NO_THROW` asserts...

File: src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp [hunk 4]
```diff
@@ -203,6 +204,185 @@ TEST(OVClassBasicTest, smoke_createMockEngineConfigThrows) {
...
+
+            is_copy_successfully = CommonTestUtils::copyFile(model_bin_name, modelBinPathW);
+            if (!is_copy_successfully) {
+                FAIL() << "Unable to copy from '" << model_bin_name << "' to '"
+                       << ::ov::util::wstring_to_string(modelBinPathW) << "'";
+            }
+
+            ov::Core core = createCoreWithTemplate();
+
+            OV_ASSERT_NO_THROW(core.compile_model(modelXmlPathW, CommonTestUtils::DEVICE_CPU,
```
Comment (ilyachur) [inline src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp:367]: Why not
```suggestion
            ASSERT_NO_THROW(core.compile_model(modelXmlPathW, CommonTestUtils::DEVICE_CPU,
```

File: src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp (hunk 3 above)
Comment (zhaixuejun1993) [reply src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp:372]: > Why do we need this `try {} catch {}`? Looks like we use `NO_THROW` asserts...

I followed the other test case style in this file. I think this `try {} catch {}` is for the other operations such as create core. If another operations throw exception we also need to clear generated test files.

File: src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp (hunk 4 above)
Comment (zhaixuejun1993) [reply src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp:367]: Seems like macro used here can provide more info for failed case, 
```
#define OV_ASSERT_NO_THROW(statement) \
  OV_ASSERT_NO_THROW_(statement, GTEST_FATAL_FAILURE_)

#define OV_ASSERT_NO_THROW_(statement, fail)    \
  GTEST_AMBIGUOUS_ELSE_BLOCKER_ \
  if (::testing::internal::AlwaysTrue()) { \
    try { \
      GTEST_SUPPRESS_UNREACHABLE_CODE_WARNING_BELOW_(statement); \
    }  catch (const std::exception& e) { \
      fail("Expected: " #statement " doesn't throw an exception.\n" \
           "  Actual: it throws.") << e.what(); \
    }  catch (...) { \
      fail("Expected: " #statement " doesn't throw an exception.\n" \
           "  Actual: it throws."); \
    } \
  }
```

Comment (ilya-lavrenov) [inline src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp:367]: why device is CPU? I suppose we need to  parametrize this test for all devices

Comment (zhaixuejun1993) [reply src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp:367]: Actually I think for all devices test should be done in `compile_model()` test without Unicode. This test is for Unicode, so hard code the device with "CPU". If we also test all devices here, maybe caused code redundancy. 
How do u think? @ilya-lavrenov

Comment (ilyachur) [reply src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp:367]: At the current moment we share this test for all plugins. So plugins tests shouldn't depend on CPU or some other device.

Comment (zhaixuejun1993) [reply src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp:367]: > moment we share this test for all plugins. So plugins tests shouldn't depend on CPU or some other device.

OK, will add it.

**Review by ilya-lavrenov — APPROVED** [review]

**Review by ilyachur — APPROVED** [review]

**Pull request merged**